Hgc l1clustring

mastrolorenzo opened the pull request:
Adding of a simple 2D and 3D clustering algorithm + cfi/cff files modified to allow the clustering parameters parsing
(Issue #85)

mastrolorenzo committed:
removal of the cluster3D dataformat in favour of a multicluster one + upgrade on the clustering algorithm + rename of some file

mastrolorenzo committed:
fix conflict

mastrolorenzo committed:
add ntuple.root

mastrolorenzo committed:
create this new branch from c2d_devel branch after rebase with PFCal-dev/hgc-tpg-devel-CMSSW_9_0_0_pre4 (include SimClustering)

jbsauvan reviewed (changes requested):
General comments:
* The content and interface of the MultiCluster class needs to be updated, at least to give access to the individual clusters. You can have a look to what is done in the ParticleFlowReco/HGCalMultiCluster class.
* The clustering implementations should be separated from the plugin algo and put in src/, similarly to what is done for the calibration.

mastrolorenzo committed:
add of HGCalClusteringImpl and HGCalMulticlusteringImpl to implement the clustering algorithms + modification to the dataformat to include a collection of clusters in the multicluster + some minor modification to the parameters python and HGCClusterAlgo.cc class

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -3,58 +3,76 @@
-      // identification variables
-      uint32_t hOverE_; 
-  };
+        edm::PtrVector<l1t::HGCalCluster>  myclusters;
```
Add trailing underscore to make clear it is a class member

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -0,0 +1,41 @@
+#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
+#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerCellCalibration.h"
+#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
+#include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
```
There are a lot of includes that doesn't seem to be needed. Remove the unnecessary includes.

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -0,0 +1,41 @@
+public:
+  
+    HGCalClusteringImpl(const edm::ParameterSet &conf);    
+    void clusterizeBase(std::unique_ptr<l1t::HGCalTriggerCellBxCollection> &, std::unique_ptr<l1t::HGCalClusterBxCollection>  &);
```
The first parameter is an "in" parameter, use a const reference to the TC collection. For the  second one, use a reference to the Cluster collection

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h:
```diff
@@ -0,0 +1,40 @@
+#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
+#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerCellCalibration.h"
+#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
+#include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
```
Clean the unnecessary includes

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h:
```diff
@@ -0,0 +1,40 @@
+public:
+
+    HGCalMulticlusteringImpl(const edm::ParameterSet &conf);    
+    void clusterizeMultiple(std::unique_ptr<l1t::HGCalClusterBxCollection> &, std::unique_ptr<l1t::HGCalMulticlusterBxCollection> &);
```
const reference to the Cluster collection and reference to the MultiCluster collection

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -0,0 +1,62 @@
+                        
+                if(trgCellLayer<28){
+                    C2d_hwPtEm+=tc->hwPt();
+                }else if(trgCellLayer>=28){
```
Does it work? I would rather check the subdet id.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -0,0 +1,83 @@
+            for(l1t::HGCalClusterBxCollection::const_iterator cl_aux = cluster_product_->begin(); cl_aux != cluster_product_->end(); ++cl_aux, ++idx){
+                //  std::cout << "     loop over Cl again and search for match:" << "   idx: " << idx << "  eta: " << cl_aux->p4().Eta() << std::endl;
+                //std::cout << "   before isMerged loop: " << skip<< std::endl;
+                for(size_t i(0); i<isMerged.size(); i++){
```
There are more efficient ways to search, e.g.:
http://en.cppreference.com/w/cpp/container/unordered_set/find

jbsauvan commented on L1Trigger/L1THGCal/test/testNtuplyzer_cfg.py:
```diff
@@ -0,0 +1,103 @@
+                              HGCalHESiliconSensitive_tag = cms.string('HGCalHESiliconSensitive'),                           
+                              calib_parameters = process.hgcalTriggerPrimitiveDigiProducer.BEConfiguration.algorithms[0].calib_parameters,
+                              seeding_threshold = process.hgcalTriggerPrimitiveDigiProducer.BEConfiguration.algorithms[0].C2d_parameters.seeding_threshold,
+                              clustering_threshold = process.hgcalTriggerPrimitiveDigiProducer.BEConfiguration.algorithms[0].C2d_parameters.clustering_threshold,
```
C2d parameters should be grouped together (as well as C3d), as done for calibration parameters

jbsauvan reviewed (commented):
In addition to the individual comments, the code will need some cleaning.

mastrolorenzo commented:
Hi JB,

I have applied almost all of your comments, but for a few of them I have 
some question. I write the question inline:

Il 22/02/2017 11:09, Jean-Baptiste Sauvan ha scritto:
>
> *@jbsauvan* commented on this pull request.
>
> In addition to the individual comments, the code will need some cleaning.
>
> ------------------------------------------------------------------------
>
> In DataFormats/L1THGCal/interface/HGCalMulticluster.h 
> <https://github.com/PFCal-dev/cmssw/pull/84#discussion_r102402826>:
>
> >      private:
> -      // Energies
> -      uint32_t hwPtEm_;
> -      uint32_t hwPtHad_;
> -      uint32_t hwSeedPt_;
> -
> -      // HGC specific information
> -      uint32_t subDet_;
> -      uint32_t nTotLayer_;
> -
> -      // identification variables
> -      uint32_t hOverE_;
> -  };
> +        edm::PtrVector<l1t::HGCalCluster>  myclusters;
>
> Add trailing underscore to make clear it is a class member
>
done,
> ------------------------------------------------------------------------
>
> In L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h 
> <https://github.com/PFCal-dev/cmssw/pull/84#discussion_r102403022>:
>
> > +#include "Geometry/HGCalCommonData/interface/HGCalDDDConstants.h"
> +
> +#include "DataFormats/ForwardDetId/interface/HGCEEDetId.h"
> +#include "DataFormats/ForwardDetId/interface/HGCHEDetId.h"
> +#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
> +#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
> +
> +
> +#include "L1Trigger/L1THGCal/interface/HGCalTriggerBackendAlgorithmBase.h"
> +#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellBestChoiceCodec.h"
> +#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellThresholdCodec.h"
> +#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
> +#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
> +#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerCellCalibration.h"
> +#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
> +#include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
>
> There are a lot of includes that doesn't seem to be needed. Remove the 
> unnecessary includes.
>
done
> ------------------------------------------------------------------------
>
> In L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h 
> <https://github.com/PFCal-dev/cmssw/pull/84#discussion_r102404799>:
>
> > +#include "L1Trigger/L1THGCal/interface/HGCalTriggerBackendAlgorithmBase.h"
> +#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellBestChoiceCodec.h"
> +#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellThresholdCodec.h"
> +#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
> +#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
> +#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerCellCalibration.h"
> +#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
> +#include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
> +
> +
> +class HGCalClusteringImpl{
> +
> +public:
> +
> +    HGCalClusteringImpl(const edm::ParameterSet &conf);
> +    void clusterizeBase(std::unique_ptr<l1t::HGCalTriggerCellBxCollection> &, std::unique_ptr<l1t::HGCalClusterBxCollection>  &);
>
> The first parameter is an "in" parameter, use a const reference to the 
> TC collection. For the second one, use a reference to the Cluster 
> collection
>
here what I did is to pass a reference to the std::pointer of 
triggerCell collection, is it what did you meant or whould I pass a 
reference directly to a const l1t::HGCalTriggerCellBxCollection? sorry 
for the confusion...

In L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h 
<https://github.com/PFCal-dev/cmssw/pull/84#discussion_r102405859>:
> > +#include "Geometry/HGCalCommonData/interface/HGCalDDDConstants.h"
> +
> +#include "DataFormats/ForwardDetId/interface/HGCEEDetId.h"
> +#include "DataFormats/ForwardDetId/interface/HGCHEDetId.h"
> +#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
> +#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
> +
> +
> +#include "L1Trigger/L1THGCal/interface/HGCalTriggerBackendAlgorithmBase.h"
> +#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellBestChoiceCodec.h"
> +#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellThresholdCodec.h"
> +#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
> +#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
> +#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerCellCalibration.h"
> +#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
> +#include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
>
> Clean the unnecessary includes
>
done,
> ------------------------------------------------------------------------
>
> In 
> L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h 
> <https://github.com/PFCal-dev/cmssw/pull/84#discussion_r102405949>:
>
> > +#include "L1Trigger/L1THGCal/interface/HGCalTriggerBackendAlgorithmBase.h"
> +#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellBestChoiceCodec.h"
> +#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellThresholdCodec.h"
> +#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
> +#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
> +#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerCellCalibration.h"
> +#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
> +#include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
> +
> +
> +class HGCalMulticlusteringImpl{
> +
> +public:
> +
> +    HGCalMulticlusteringImpl(const edm::ParameterSet &conf);
> +    void clusterizeMultiple(std::unique_ptr<l1t::HGCalClusterBxCollection> &, std::unique_ptr<l1t::HGCalMulticlusterBxCollection> &);
>
> const reference to the Cluster collection and reference to the 
> MultiCluster collection
>
simular doubt as for the comment on the HGCalClusterImpl.h regarding the 
input collection

In L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc 
<https://github.com/PFCal-dev/cmssw/pull/84#discussion_r102406720>:
> > +        uint32_t C2d_hwPtEm = 0;
> +        uint32_t C2d_hwPtHad = 0;
> +
> +//        assert trgcell_product_->size() > 0 //&& put logWarning
> +
> +        for(l1t::HGCalTriggerCellBxCollection::const_iterator tc = trgcell_product_->begin(); tc != trgcell_product_->end(); ++tc)
> +        {
> +            if(tc->hwPt()>0)
> +            {
> +
> +                HGCalDetId trgdetid(tc->detId());
> +                int trgCellLayer = trgdetid.layer();
> +
> +                if(trgCellLayer<28){
> +                    C2d_hwPtEm+=tc->hwPt();
> +                }else if(trgCellLayer>=28){
>
> Does it work? I would rather check the subdet id.
>
why did you think the trgdetid.layer() is not working? maybe I missed 
something...
> ------------------------------------------------------------------------
>
> In L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc 
> <https://github.com/PFCal-dev/cmssw/pull/84#discussion_r102410724>:
>
> > +            double_t C3d_pt  = 0.;
> +            double_t C3d_eta = 0.;
> +            double_t C3d_phi = 0.;
> +            uint32_t C3d_hwPtEm = 0;
> +            uint32_t C3d_hwPtHad = 0;
> +            uint32_t totLayer = 0;
> +
> +            bool skip=false;
> +
> +            //std::cout << "In the Cl collection, seed the C3d with this : " << seedx << " - "<< cl->p4().Pt() << " eta: " <<  cl->p4().Eta() << " --> layer " << cl->layer() << "  skip before 2nd loop "<< skip << std::endl;
> +
> +            size_t idx=0;
> +            for(l1t::HGCalClusterBxCollection::const_iterator cl_aux = cluster_product_->begin(); cl_aux != cluster_product_->end(); ++cl_aux, ++idx){
> +                //  std::cout << "     loop over Cl again and search for match:" << "   idx: " << idx << "  eta: " << cl_aux->p4().Eta() << std::endl;
> +                //std::cout << "   before isMerged loop: " << skip<< std::endl;
> +                for(size_t i(0); i<isMerged.size(); i++){
>
> There are more efficient ways to search, e.g.:
> http://en.cppreference.com/w/cpp/container/unordered_set/find
>
thank! However I didn't modify it in this commit since I am going 
completely to change the 2D algorithm (putting first a simple seed+ 
geometric-C2d and then the NN one) so I will use this method later maybe.
> ------------------------------------------------------------------------
>
> In L1Trigger/L1THGCal/test/testNtuplyzer_cfg.py 
> <https://github.com/PFCal-dev/cmssw/pull/84#discussion_r102436465>:
>
> > +
> +
> +trgCells_algo_all =  cms.PSet( AlgorithmName = cms.string('SingleCellClusterAlgoBestChoice'),
> +                              FECodec = process.hgcalTriggerPrimitiveDigiProducer.FECodec,
> +                              HGCalEESensitive_tag = cms.string('HGCalEESensitive'),
> +                              HGCalHESiliconSensitive_tag = cms.string('HGCalHESiliconSensitive'),
> +
> +                              calib_parameters = process.hgcalTriggerPrimitiveDigiProducer.BEConfiguration.algorithms[0].calib_parameters
> +                              )
> +cluster_algo_all =  cms.PSet( AlgorithmName = cms.string('HGCClusterAlgoBestChoice'),
> +                              FECodec = process.hgcalTriggerPrimitiveDigiProducer.FECodec,
> +                              HGCalEESensitive_tag = cms.string('HGCalEESensitive'),
> +                              HGCalHESiliconSensitive_tag = cms.string('HGCalHESiliconSensitive'),
> +                              calib_parameters = process.hgcalTriggerPrimitiveDigiProducer.BEConfiguration.algorithms[0].calib_parameters,
> +                              seeding_threshold = process.hgcalTriggerPrimitiveDigiProducer.BEConfiguration.algorithms[0].C2d_parameters.seeding_threshold,
> +                              clustering_threshold = process.hgcalTriggerPrimitiveDigiProducer.BEConfiguration.algorithms[0].C2d_parameters.clustering_threshold,
>
> C2d parameters should be grouped together (as well as C3d), as done 
> for calibration parameters
>
> —
>
done,

> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/PFCal-dev/cmssw/pull/84#pullrequestreview-23155684>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFS0Z31GkJx1i2tNAiAyznW_PxSj6O61ks5rfBdQgaJpZM4MC8_9>.
>
let me know if do you have any further comments or suggestion.
Thank you very much,

cheers,
Luca


-- 
---------------------------------------------------------------
Dr. Luca Mastrolorenzo
CERN - CMS experiment
Blackett Laboratory - Imperial College London
Prince Consort Rd, London SW7 2BB
United Kingdom

Centre Européen pour la Recherche Nucléaire
Route de Meyrin 385, 1217 Meyrin, Suisse

Office: 40-4B-08
Téléphone: +44 777 8 224 851
---------------------------------------------------------------

jbsauvan commented:
Hi @mastrolorenzo
 > or whould I pass a reference directly to a const l1t::HGCalTriggerCellBxCollection?

Yes. const l1t::HGCalTriggerCellBxCollection&. And similarly with clusters.

> why did you think the trgdetid.layer() is not working?

Because FH layer numbers are <= 13

mastrolorenzo commented:
Ah yes sure, it was just an example, it worked however fine, but that piece is going to be rewritten
Thanks!
Cheers 
Luca

Inviato da iPhone

> Il giorno 23 feb 2017, alle ore 17:34, Jean-Baptiste Sauvan <notifications@github.com> ha scritto:
> 
> Hi @mastrolorenzo
> 
> or whould I pass a reference directly to a const l1t::HGCalTriggerCellBxCollection?
> 
> Yes. const l1t::HGCalTriggerCellBxCollection&. And similarly with clusters.
> 
> why did you think the trgdetid.layer() is not working?
> 
> Because FH layer numbers are <= 13
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mastrolorenzo committed:
this is the version of the code using pointers in the context of the clustering and multiclutering next commit will use references

mastrolorenzo committed:
modifications to the cluster implementation functions in order to take collection references instead of unique_ptr to collections + add of parameter for geometrical cluster 2D in hgcalTriggerPrimitiveDigiProducer_cfi.py

mastrolorenzo committed:
c2d clustering algorithm core is compiling: defined addTC() and dist() member functions as const in HGCalCluster class

mastrolorenzo committed:
2D cluster is ready to be tested

mastrolorenzo committed:
small modification to the 2D-clustering algorithm + 3D-clustering as in IC code just to debug the C2d (need the sum of the energy)

mastrolorenzo committed:
2D cluster algorithm implemented (simple 2D-dR(x,y) approach)
+ modifications to the cluster classes and their implementation:
parsing of the eventSetup and Event in the clusterise functions
(needed to retrieve GlobalPointers to the cartesian position of the 2D) + creation of a *temporary-C3d* used to test the C2d + global cleaning

jbsauvan commented:
Closing this PR
The first set of commits has been merged with PR #86 
The rest (to be rebased) will go in an other PR
